crates/ra_hir/src/nameres/collector.rs

```diff
@@ -295,6 +295,23 @@ where
                         }
                     }
 
+                    // `#[macro_use] extern crate` glob import macros
+                    if import.is_extern_crate && import.is_macro_use {
+                        if let Some(ModuleDef::Module(m)) =
+                            def.a().and_then(|item| item.take_types())
+                        {
+                            let item_map = self.db.crate_def_map(m.krate);
+                            let scope = &item_map[m.module_id].scope;
+                            let macros = scope
+                                .macros
+                                .iter()
+                                .map(|(name, res)| (name.clone(), Either::B(*res)))
+                                .collect::<Vec<_>>();
+
+                            self.update(module_id, Some(import_id), &macros);
+                        }
+                    }
```

Comment:
This I think goes into a slightly wrong direction at the moment. 

The problem is that there are essentially two modes for resolving macros.

In the new mode, macros are items, and are imported by paths.

In the old mode, there's a mutable set of currently defined macros in the parser, and `foo!()` invocation is looked up in the map.

This code uses the new model of scopes, but macro_use / macro_export are fundamentally tied to the old mode.

Instead, I think we should do the following:

* Add `exported_macros: FxHashSet<Name, MacroDef>` to `CrateDefMap` (note, this should be crate level)
* Populate this set when we collect the crate
* When we encounter an `#[macro_use] extern crate`, add the crates `exported_macros` to `global_macro_scope` in `DeCollector`

Reply:
Also, let's put a [mark](https://github.com/rust-analyzer/rust-analyzer/blob/f2a200c1ee8016f47b35e2e13a899fc96286a1eb/crates/test_utils/src/marks.rs) into this `if` and the corresponding test

Reply:
And I think we need some special handing for `core` and `std` crates, which are `macro_use` even without an explicit `extern crate`

Reply:
Fixed. 

For the prelude, it is now handled in name resolving. Would it be better to also search `exported_macros` when resolving names ([here](https://github.com/rust-analyzer/rust-analyzer/blob/a66214b34effe1ad7f4351a1b920cf3a8f98d3c0/crates/ra_hir/src/nameres.rs#L506)) instead of early-import?